feat(Topology/Algebra/InfiniteSum): Discrete convolution and Cauchy product APIs

IlPreteRosso opened the pull request:
This is the first PR is a series of PRs towards the proposed API and ℓ1 Banach algebra. 

Defines discrete convolution of functions `f : M → E` and `g : M → E'` over a monoid `M`
using a bilinear map `L : E →ₗ[S] E' →ₗ[S] F`:

  `(f ⋆[L] g) x = ∑' (a, b) : mulFiber x, L (f a) (g b)`
  
  where `def mulFiber (x : M) : Set (M × M) := Set.mulAntidiagonal Set.univ Set.univ x`
  
  
  The major changes are (from the last review)
- Move file to new location
- Change protected to private, add irreducible attribute to `tripleFiber` (~25 lines)
- Add convolution assoc variant theorems (see RM below) (~100 lines)
- Add distribution theorems and smul/identity related lemmas (~50 lines)
- Finish part of TODO in [Prod.lean](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/Antidiag/Prod.html#Finset.sigmaAntidiagonalEquivProd) to match API needs. Basically restating the defs and theorems in @[to_additive]. (~50 lines)
- Also removed some less useful lemmas: some linearMap.Mul alias of bilinear version lemmas by just passing in linearMap.Mul ℕ R; remove convolution_apply, just use the definition of convolution; 

RM: While `ringConvolution_assoc` (reviewed last time) is preparing for Cauchy product (Finset, see section MulAntidiagonal), the new `completeUniformRingConvolution_assoc` is for lpOneBanachAlgebra (tsum, complete space, less compatibility hypotheses needed), which is going to be another PR.

IlPreteRosso committed:
fresh commit

IlPreteRosso committed:
lake exe mk_all

IlPreteRosso committed:
fix lint

IlPreteRosso committed:
lake exe mk_all

IlPreteRosso committed:
cleanup

IlPreteRosso committed:
protected def mulMap

IlPreteRosso committed:
Set.mulAntidiagonal

IlPreteRosso committed:
tripleFiber as mulTripleAntidiagonal

IlPreteRosso committed:
remove redundant simp from to_additive

IlPreteRosso committed:
add mulFiber_swapEquiv

IlPreteRosso committed:
improve doc

IlPreteRosso committed:
ConvolutionExistsAt

IlPreteRosso committed:
clean

IlPreteRosso committed:
change attr=simp to simp

IlPreteRosso committed:
🧹

IlPreteRosso committed:
Retrigger CI

IlPreteRosso committed:
Retrigger CI

IlPreteRosso committed:
clean

j-loreaux commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,277 @@
+  /-- The fiber of triple addition at `x`: all triples `(a, b, c)` with `a + b + c = x`.
+  This is `addTripleAntidiagonal Set.univ Set.univ Set.univ x`. -/]
+abbrev tripleFiber (x : M) : Set (M × M × M) :=
+  mulTripleAntidiagonal Set.univ Set.univ Set.univ x
```
It's not clear to me why this is necessary. Couldn't you just have a `assocEquiv` between the sigma types (i.e., just compose `leftAssocEquiv` with `rightAssocEquiv.symm`)?

IlPreteRosso commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,277 @@
+# Discrete Convolution
+
+Discrete convolution of `f : M → E` and `g : M → E'` over a monoid `M`:
+`(f ⋆[L] g) x = ∑' (a, b) : mulFiber x, L (f a) (g b)` where `mulFiber x = {(a, b) | a * b = x}`.
```
In the local working file (will come last in this series of PRs) I have result
```
theorem addMulConvolution_eq_measureTheory_convolution [Countable M]
    (f g : M → R) (hfg : ∀ x, Integrable (fun t => f t * g (x - t)) .count) :
    (f ⋆₊ₘ g) = MeasureTheory.convolution f g (ContinuousLinearMap.mul ℝ R) .count := by
```

I will add more docstrings to the next PR from local projects.

IlPreteRosso committed:
Remove sigma, add assocEquiv, convolution_assoc, ringConvolution_assoc

IlPreteRosso committed:
remove Main Definitions from docstring

IlPreteRosso committed:
linter

IlPreteRosso committed:
clean

IlPreteRosso committed:
improve doc

j-loreaux commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,488 @@
+  /-- `addTripleAntidiagonal s t u a` is the set of all triples `(x, y, z)` with `x ∈ s`, `y ∈ t`,
+  `z ∈ u`, and `x + y + z = a`. Triple analog of `Set.addAntidiagonal`. -/]
+protected def mulTripleAntidiagonal (s t u : Set M) (a : M) : Set (M × M × M) :=
+  {x | x.1 ∈ s ∧ x.2.1 ∈ t ∧ x.2.2 ∈ u ∧ x.1 * x.2.1 * x.2.2 = a}
```
It's fine to have this, but I think you intended for it to be `private`, not `protected`.

IlPreteRosso replied:
fixed. Now leaving only the following theorem public to bridge between `convolution_assoc` summable hypothesis and private implementations. 
```
theorem mem_tripleFiber {x : M} {abc : M × M × M} :
    abc ∈ tripleFiber x ↔ abc.1 * abc.2.1 * abc.2.2 = x :
```

IlPreteRosso committed:
add missing to_additive

IlPreteRosso committed:
clean

IlPreteRosso committed:
clean

j-loreaux commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,485 @@
+# Discrete Convolution
+
+Discrete convolution over monoids: `(f ⋆[L] g) x = ∑' (a, b) : mulFiber x, L (f a) (g b)`
+where `mulFiber x = {(a, b) | a * b = x}`.
```
```suggestion
Discrete convolution over monoids: `(f ⋆[L] g) x = ∑' (a, b) : mulFiber x, L (f a) (g b)`
where `mulFiber x = {(a, b) | a * b = x}`. Additive monoids are also supported.
```

j-loreaux commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,485 @@
+
+## Design
+
+Uses bilinear map `L : E →ₗ[S] E' →ₗ[S] F` to combine values, following `MeasureTheory.convolution`.
```
```suggestion
Uses a bilinear map `L : E →ₗ[S] E' →ₗ[S] F` to combine values, following `MeasureTheory.convolution`.
```

j-loreaux commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,485 @@
+Uses bilinear map `L : E →ₗ[S] E' →ₗ[S] F` to combine values, following `MeasureTheory.convolution`.
+For specializing to ring multiplication, use `ringConvolution` = `convolution (LinearMap.mul R R)`.
+
+Index monoid `M` can be non-commutative (group algebras R[G] with non-abelian G).
```
```suggestion
Then index monoid `M` can be non-commutative (group algebras R[G] with non-abelian G).
```

IlPreteRosso committed:
Make tripleFiber implementation private. `mem_tripleFiber` is public as a part of the API

IlPreteRosso committed:
Update Mathlib/Analysis/DiscreteConvolution/Basic.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

IlPreteRosso committed:
Update Mathlib/Analysis/DiscreteConvolution/Basic.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

IlPreteRosso committed:
Update Mathlib/Analysis/DiscreteConvolution/Basic.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

IlPreteRosso committed:
Update Mathlib/Analysis/DiscreteConvolution/Basic.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

IlPreteRosso committed:
Update Mathlib/Analysis/DiscreteConvolution/Basic.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

IlPreteRosso committed:
docstring

IlPreteRosso commented on Mathlib/Analysis/DiscreteConvolution/Basic.lean:
```diff
@@ -0,0 +1,485 @@
+/-- Convolution using ring multiplication. This is `convolution (LinearMap.mul R R)`. -/
+@[to_additive (dont_translate := R) addRingConvolution
+  /-- Additive convolution using ring multiplication. -/]
+def ringConvolution (f g : M → R) : M → R := convolution (LinearMap.mul R R) f g
```
Good point. Fixed.

IlPreteRosso committed:
ℕ R instead of R R

IlPreteRosso committed:
lint timeout

IlPreteRosso committed:
docstring

IlPreteRosso committed:
more private and docstring

IlPreteRosso committed:
more docstring

IlPreteRosso committed:
AntidiagonalRing, CauchyProduct, relocate files.

IlPreteRosso committed:
Mathlib.lean

IlPreteRosso committed:
docstring

IlPreteRosso committed:
IsTopologicalSemiring

IlPreteRosso committed:
docstring

IlPreteRosso committed:
rename completeUniformRingConvolution_assoc_at

IlPreteRosso committed:
remove too restrictive assoc variant

IlPreteRosso committed:
docstring

IlPreteRosso committed:
Finish part of TODO in /Users/ilpreterosso/VSCode/LEAN/Mathlib/mathlib4/Mathlib/Algebra/Order/Antidiag/Prod.lean

added HasMulAntidiagonal, which converts to the existing HasAntidiagonal via to_additive.

Update CauchyProduct to use HasMulAntidiagonal and to_additive

IlPreteRosso committed:
refactor

IlPreteRosso committed:
docstring

IlPreteRosso committed:
docstring

IlPreteRosso committed:
add open scoped CauchyProduct for better signature display

IlPreteRosso committed:
update some dont_translate

IlPreteRosso committed:
update section MulAntidiagonal according to HasMulAntidiagonal

IlPreteRosso committed:
everything is to_additive

IlPreteRosso commented:
The mathlib lint is acting stupid can’t pass check

IlPreteRosso commented:
> You have a build error you should take care of first

I'm on it 😆. It is a historical issue left in the TODO of Prod.lean, some naming duplication got triggered when refactoring to using to_additive

IlPreteRosso committed:
avoid name collision with the existing Finset.swap_mem_mulAntidiagonal in MulAntidiagonal.lean

IlPreteRosso committed:
clean

IlPreteRosso commented:
Add Finset.Antidiag and Finset.MulAntidiag namespaces 

Resolves naming collision between [Finset.swap_mem_mulAntidiagonal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finset/MulAntidiagonal.html#Finset.swap_mem_mulAntidiagonal) in MulAntidiagonal.lean and Prod.lean

IlPreteRosso committed:
section name

IlPreteRosso committed:
improve notation API and avoid notation conflict

IlPreteRosso committed:
clean

IlPreteRosso committed:
clean

IlPreteRosso committed:
remove simp attribute from pivate defs/theorems; add (attr := irreducible) to `addAssocEquiv`

IlPreteRosso committed:
set assocEquiv back to private

IlPreteRosso committed:
remove redundant bridge theorems

IlPreteRosso committed:
remove repeated section

IlPreteRosso committed:
clean

IlPreteRosso committed:
docstring and notation

IlPreteRosso committed:
Retrigger CI

IlPreteRosso committed:
default notations to right associativity as in measure theory convolution API

IlPreteRosso committed:
Update notation precedence to exactly as in measure theory convolution

IlPreteRosso committed:
clean

IlPreteRosso committed:
clean+

IlPreteRosso committed:
clean++

IlPreteRosso committed:
clean pro

IlPreteRosso committed:
add distribution theorems

IlPreteRosso committed:
clean

IlPreteRosso committed:
refacor ang smul lemmas

IlPreteRosso committed:
add identity properties

IlPreteRosso committed:
clean

IlPreteRosso committed:
oopsy

IlPreteRosso committed:
remove CauchyProduct for another PR

IlPreteRosso committed:
clean

IlPreteRosso committed:
remove some ring convolution lemmas

IlPreteRosso commented:
The major changes are (from the last review)
- Move file to new location
- Change protected to private, add irreducible attribute (~25 lines)
- Add convolution assoc variant theorems (see RM below) (~150 lines)
- Add distribution theorems and smul/identity related lemmas (~100 lines)
- Add CauchyProduct API (lightweight, standalone, with bridge to discrete convolution API) (~200 lines) (**removed for brevity**)
- Finish part of TODO in [Prod.lean](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/Antidiag/Prod.html#Finset.sigmaAntidiagonalEquivProd) to match API needs. Basically restating the defs and theorems in @[to_additive]. (~50 lines)

RM: While `ringConvolution_assoc` (reviewed last time) is preparing for Cauchy product (Finset, see section CauchyProductBridge), the new `completeUniformRingConvolution_assoc` is for lpOneBanachAlgebra (tsum), which is going to be another PR. 

The extra length is more of less for user friendly API design, variants of previous theorems, and some refactoring using to_additive. Not too much brand new things.

IlPreteRosso committed:
private

IlPreteRosso committed:
docstring

IlPreteRosso committed:
Merge branch 'leanprover-community:master' into Discrete-Convolution-API-Basics

IlPreteRosso committed:
Merge branch 'leanprover-community:master' into Discrete-Convolution-API-Basics
